presto-hive/src/main/java/com/facebook/presto/hive/util/InternalHiveSplitFactory.java

```diff
@@ -186,8 +181,7 @@ public InternalHiveSplitFactory(
                 readBucketNumber,
                 tableBucketNumber,
                 splittable,
-                forceLocalScheduling && allBlocksHaveRealAddress(blocks),
-                (nodeSelectionStrategy == HARD_AFFINITY && !allBlocksHaveRealAddress(blocks)) ? NO_PREFERENCE : nodeSelectionStrategy,
+                forceLocalScheduling && allBlocksHaveRealAddress(blocks) ? HARD_AFFINITY : nodeSelectionStrategy,
```

Comment:
You need `&& nodeSelectionStrategy == HARD_AFFINITY`

Reply:
> You need `&& nodeSelectionStrategy == HARD_AFFINITY`

IDE indicates that It's actually redundant and it is